### PR TITLE
Provide blockie identicon option

### LINKED ISSUE #1785: Add Blockie icon option
Let users opt into blockies if they want.

Both parity and etherscan use https://github.com/ethereum/blockies

### PR DESCRIPTION
**Pull request by ghost**
Fixes #1785

### CODE REVIEW

**Comment by jpclark** [comment]
The tests that are failing are not related to the changes I made and seem to be there from before.

File: .eslintignore [hunk 1]
```diff
@@ -2,4 +2,5 @@ app/scripts/lib/extension-instance.js
 test/integration/bundle.js
 test/integration/jquery-3.1.0.min.js
 test/integration/helpers.js
-test/integration/lib/first-time.js
\ No newline at end of file
+test/integration/lib/first-time.js
+ui/lib/blockies.js
```
Comment (danfinlay) [inline .eslintignore]: Is it necessary to ignore your new file in the linting?

File: app/scripts/controllers/preferences.js [hunk 2]
```diff
@@ -14,6 +14,14 @@ class PreferencesController {
   }
 // PUBLIC METHODS
 
+  setUseBlockie (val) {
+    this.store.updateState({ useBlockie: val })
```
Comment (danfinlay) [inline app/scripts/controllers/preferences.js]: Should also include a "default value" for this key in the constructor.

File: package.json [hunk 3]
```diff
@@ -100,8 +100,7 @@
     "extensionizer": "^1.0.0",
     "fast-json-patch": "^2.0.4",
     "fast-levenshtein": "^2.0.6",
-    "fuse.js": "^3.1.0",
-    "gulp": "github:gulpjs/gulp#4.0",
```
Comment (danfinlay) [inline package.json]: I'd usually ask for dependency cleanups to be in separate PRs, so they don't mire up discussion over a proposal.

File: ui/app/components/identicon.js [hunk 4]
```diff
@@ -16,16 +18,22 @@ function IdenticonComponent () {
...
-  const { className = '', address } = props
+  const { className = '', address, useBlockie } = props
   var diameter = props.diameter || this.defaultDiameter
 
   return address
     ? (
       h('div', {
         className: `${className} identicon`,
-        key: 'identicon-' + address,
+        key: useBlockie ? 'blockie' : 'identicon-' + address,
```
Comment (danfinlay) [inline ui/app/components/identicon.js]: `Key` needs to be unique to the element in case it's used in array, so including the address is a requirement, even if a blockie.

File: ui/app/components/identicon.js [hunk 5]
```diff
@@ -51,38 +59,59 @@ IdenticonComponent.prototype.render = function () {
...
   if (!isNode) {
-    var img = iconFactory.iconForAddress(address, diameter)
-    container.appendChild(img)
+    // eslint-disable-next-line react/no-find-dom-node
+    var container = findDOMNode(this)
+
+    const diameter = props.diameter || this.defaultDiameter
+
+    if (useBlockie) {
+      _generateBlockie(container, address)
```
Comment (danfinlay) [inline ui/app/components/identicon.js]: Should include `diameter` parameter here, right?

**Review by danfinlay — CHANGES_REQUESTED** [review]
This is a fantastic achievement, thank you so much for doing it. It's not often that someone enters the project and makes changes they want so easily, with so few questions, so a strong congratulations.

There are a few questions and semantical points that are worth addressing before merging, and I hope they don't discourage you. I'm very happy to see a fresh contributor, and I hope to put up more, healthier bounties for people like you to pursue in the future.

File: .eslintignore (hunk 1 above)
Comment (jpclark) [reply .eslintignore]: Excellent question. This is the lib file from the MyEtherWallet project. I originally did not leave this file out of the eslint checks but it was blowing up Lint with a ton of styling preference "errors" (dangling commas etc). Apparently the MEW folks didn't follow the same style guide ;) I was faced with a dilemma: edit a library containing trusted code to fix some style issues, or omit the file from lint checking. I chose to leave it out of the checks.

File: app/scripts/controllers/preferences.js (hunk 2 above)
Comment (jpclark) [reply app/scripts/controllers/preferences.js]: I'm probably just being stupid, but I'm not sure what you mean. The useBlockie property is initialized with a false value by the metamask.js state before Redux syncs the prop values up, is this not sufficient?

File: package.json (hunk 3 above)
Comment (jpclark) [reply package.json]: Oooh, yeah sorry I didn't catch that in my commit diffs. I must have accidentally updated the project deps with NPM when I was setting up the project. I can revert those changes.

File: ui/app/components/identicon.js (hunk 4 above)
Comment (jpclark) [reply ui/app/components/identicon.js]: Gotcha, fixing that now.

File: ui/app/components/identicon.js (hunk 5 above)
Comment (jpclark) [reply ui/app/components/identicon.js]: Yes, yes indeed. Fixing.

**Comment by jpclark** [comment]
@danfinlay I've cleaned things up a little. Let me know what you think.

**Review by tmashuang — APPROVED** [review]
Looks good adding a gif for reference for this enhancement.
block-toggle-setting

**Comment by danfinlay** [comment]
Yay! One of our first bounties ever posted, now filled!

**Comment by cjeria** [comment]
@tmashuang @danfinlay this looks good! the blockies inside of the dropdown though look squared. Probably should make them rounded like we do currently for consistency

image
